File: rust/src/service.rs
use inapi::{Host, Package, PackageResult, Service, ServiceRunnable, Telemetry};
use std::process::Command;
use super::Testable;

pub struct ServiceTest;

impl Testable for ServiceTest {
    fn test(mut host: &mut Host) {
        let telemetry = Telemetry::init(&mut host).unwrap();

        if telemetry.os.platform == "centos" {
            let mut pkg_epel = Package::new(&mut host, "epel-release", None).unwrap();
            if let PackageResult::Result(cmd) = pkg_epel.install(&mut host).unwrap() {
                assert_eq!(cmd.exit_code, 0);
            }
        }

        let mut pkg = Package::new(&mut host, "nginx", None).unwrap();
        if let PackageResult::Result(cmd) = pkg.install(&mut host).unwrap() {
            assert_eq!(cmd.exit_code, 0);
        }

        let service = Service::new_service(ServiceRunnable::Service("nginx"), None);

        let enable = service.action(&mut host, "enable").unwrap();
        assert_eq!(enable.exit_code, 0);
        match telemetry.os.platform.as_ref() {
            "centos" => {
                let cmd = Command::new("bash").args(&["-c", "chkconfig|egrep -qs 'nginx.+3:on'"]).output().unwrap();
                assert!(cmd.status.success());
            },
            "fedora" => {
                let cmd = Command::new("bash").args(&["-c", "systemctl list-unit-files|egrep -qs 'nginx.service.+enabled'"]).output().unwrap();
                assert!(cmd.status.success());
            },
            "debian" | "ubuntu" => {
                let cmd = Command::new("bash").args(&["-c", "ls /etc/rc3.d|egrep -qs 'S[0-9]+nginx'"]).output().unwrap();
                assert!(cmd.status.success());
            },
            "freebsd" => {
                let cmd = Command::new("grep").args(&["-qs", "nginx_enable=\"YES\"", "/etc/rc.conf"]).output().unwrap();
                assert!(cmd.status.success());
            }
            _ => unimplemented!()
        }

        let start = service.action(&mut host, "start").unwrap();
        assert_eq!(start.exit_code, 0);
        let start_cmd = Command::new("pgrep").arg("nginx").output().unwrap();
        assert!(start_cmd.status.success());

        let stop = service.action(&mut host, "stop").unwrap();
        assert_eq!(stop.exit_code, 0);
        let stop_cmd = Command::new("pgrep").arg("nginx").output().unwrap();
        assert_eq!(stop_cmd.status.code().unwrap(), 1);

        let disable = service.action(&mut host, "disable").unwrap();
        assert_eq!(disable.exit_code, 0);
        match telemetry.os.platform.as_ref() {
            "centos" => {
                let cmd = Command::new("bash").args(&vec!["-c", "chkconfig|egrep -qs 'nginx.+3:off'"]).output().unwrap();
                assert!(cmd.status.success());
            },
            "fedora" => {
                let cmd = Command::new("bash").args(&vec!["-c", "systemctl list-unit-files|egrep -qs 'nginx.service.+disabled'"]).output().unwrap();
                assert!(cmd.status.success());
            },
            "debian" | "ubuntu" => {
                let cmd = Command::new("bash").args(&vec!["-c", "ls /etc/rc3.d|egrep -qs 'S[0-9]+nginx'"]).output().unwrap();
                assert!(cmd.status.success() == false);
            },
            "freebsd" => {
                let cmd = Command::new("grep").args(&vec!["-qs", "nginx_enable", "/etc/rc.conf"]).output().unwrap();
                assert!(cmd.status.success() == false);
            }
            _ => unimplemented!()
        }

        let mut pkg = Package::new(&mut host, "nginx", None).unwrap();
        if let PackageResult::Result(cmd) = pkg.uninstall(&mut host).unwrap() {
            assert_eq!(cmd.exit_code, 0);
        }
    }
}

File: php/src/service.php
<?php

use Intecture\Package;
use Intecture\Service;
use Intecture\ServiceRunnable;
use Intecture\Telemetry;

class ServiceTest implements Testable {
    public static function test($host) {
        $telemetry = Telemetry::load($host);
        $tdata = $telemetry->get();

        if ($tdata['os']['platform'] == 'centos') {
            $epel_pkg = new Package($host, 'epel-release');
            $result = $epel_pkg->install($host);
            if ($result !== NULL) {
                assert($result['exit_code'] == 0);
            }
        }

        $pkg = new Package($host, 'nginx');
        $result = $pkg->install($host);
        if ($result !== NULL) {
            assert($result['exit_code'] == 0);
        }

        $runnable = new ServiceRunnable('nginx', ServiceRunnable::SERVICE);
        $svc = new Service($runnable, array('map-enable' => 'enable'));

        $result = $svc->action($host, 'map-enable');
        if ($result !== NULL) {
            assert($result['exit_code'] == 0);
        }
        $cmd = '';
        $out = $exit = NULL;
        switch ($tdata['os']['platform']) {
            case 'centos':
                $cmd = "chkconfig|egrep -qs 'nginx.+3:on'";
                break;
            case 'fedora':
                $cmd = "systemctl list-unit-files|egrep -qs 'nginx.service.+enabled'";
                break;
            case 'debian':
            case 'ubuntu':
                $cmd = "ls /etc/rc3.d|egrep -qs 'S[0-9]+nginx'";
                break;
            case 'freebsd':
                $cmd = 'grep -qs \'nginx_enable="YES"\' /etc/rc.conf';
                break;
            default:
                fwrite(STDERR, 'Unknown platform');
                exit(1);
        }
        $check_enable = exec($cmd, $out, $exit);
        assert($exit == 0);

        assert($svc->action($host, 'map-enable') === NULL);

        $result = $svc->action($host, 'start');
        if ($result !== NULL) {
            assert($result['exit_code'] == 0);
        }
        $out = $exit = NULL;
        $check_start = exec('pgrep nginx', $out, $exit);
        assert($exit == 0);

        assert($svc->action($host, 'start') === NULL);

        $result = $svc->action($host, 'stop');
        assert($result['exit_code'] == 0);
        $out = $exit = NULL;
        $check_start = exec('pgrep nginx', $out, $exit);
        assert($exit == 1);

        assert($svc->action($host, 'stop') === NULL);

        $result = $svc->action($host, 'disable');
        assert($result['exit_code'] == 0);
        $cmd = '';
        $out = $exit = NULL;
        $expect = 0;
        switch ($tdata['os']['platform']) {
            case 'centos':
                $cmd = "chkconfig|egrep -qs 'nginx.+3:off'";
                break;
            case 'fedora':
                $cmd = "systemctl list-unit-files|egrep -qs 'nginx.service.+disabled'";
                break;
            case 'debian':
            case 'ubuntu':
                $cmd = "ls /etc/rc3.d|egrep -qs 'S[0-9]+nginx'";
                $expect = 1;
                break;
            case 'freebsd':
                $cmd = 'grep -qs \'nginx_enable="YES"\' /etc/rc.conf';
                $expect = 1;
                break;
            default:
                fwrite(STDERR, 'Unknown platform');
                exit(1);
        }
        $check_enable = exec($cmd, $out, $exit);
        assert($exit == $expect);

        assert($svc->action($host, 'disable') === NULL);

        $result = $pkg->uninstall($host);
        assert($result['exit_code'] == 0);
    }
}
